packages/@react-aria/textfield/src/useFormattedTextField.ts

```diff
@@ -0,0 +1,152 @@
+      let nextValue: string;
+      switch (e.inputType) {
+        case 'historyUndo':
+        case 'historyRedo':
+          // Explicitly allow undo/redo. e.data is null in this case, but there's no need to validate,
+          // because presumably the input would have already been validated previously.
+          return;
+        case 'deleteContent':
+        case 'deleteByCut':
+        case 'deleteByDrag':
+          nextValue = input.value.slice(0, input.selectionStart) + input.value.slice(input.selectionEnd);
+          break;
+        case 'deleteContentForward':
+          // This is potentially incorrect, since the browser may actually delete more than a single UTF-16
+          // character. In reality, a full Unicode grapheme cluster consisting of multiple UTF-16 characters
+          // or code points may be deleted. However, in our currently supported locales, there are no such cases.
+          // If we support additional locales in the future, this may need to change.
+          nextValue = input.selectionEnd === input.selectionStart
+            ? input.value.slice(0, input.selectionStart) + input.value.slice(input.selectionEnd + 1)
+            : input.value.slice(0, input.selectionStart) + input.value.slice(input.selectionEnd);
+          break;
+        case 'deleteContentBackward':
+          nextValue = input.selectionEnd === input.selectionStart
+            ? input.value.slice(0, input.selectionStart - 1) + input.value.slice(input.selectionStart)
+            : input.value.slice(0, input.selectionStart) + input.value.slice(input.selectionEnd);
+          break;
+        default:
+          if (e.data != null) {
+            nextValue =
+              input.value.slice(0, input.selectionStart) +
+              e.data +
+              input.value.slice(input.selectionEnd);
+          }
+          break;
+      }
+
+      // If we did not compute a value, or the new value is invalid, prevent the event
+      // so that the browser does not update the input text, move the selection, or add to
+      // the undo/redo stack.
+      if (nextValue == null || !state.validate(nextValue)) {
+        e.preventDefault();
+      }
+    };
+
+    input.addEventListener('beforeinput', onBeforeInput, false);
+    return () => {
+      input.removeEventListener('beforeinput', onBeforeInput, false);
+    };
+  }, [inputRef, stateRef]);
+
+  let onBeforeInput = !supportsNativeBeforeInputEvent()
+    ? e => {
+      let nextValue =
+        e.target.value.slice(0, e.target.selectionStart) +
+        e.data +
+        e.target.value.slice(e.target.selectionEnd);
+
+      if (!state.validate(nextValue)) {
+        e.preventDefault();
+      }
+    }
+    : null;
+
+  let {labelProps, inputProps: textFieldProps} = useTextField(props, inputRef);
+
+  let compositionStartState = useRef(null);
+  return {
+    inputProps: {
+      ...textFieldProps,
```

Comment:
do we need to use `mergeProps` here? user might have their own `onBeforeInput` etc.

Reply:
ah, that's a good point